### ISSUE TITLE
truncated bptt without padding?

### ISSUE THREAD

**Issue by eric-haibin-lin** [Stonesjtu/Pytorch-NCE#2]
Hi,

Thanks for the great example. I noticed that you pad sentences to the max length per mini-batch, which is a bit different from the truncated bptt approach of the original [word_language_model without NCE](https://github.com/pytorch/examples/tree/master/word_language_model) on pytorch. I wonder if you have compared the two approach and investigated if it makes a difference in terms of the final ppl?

I'm also interested to know how much better this model can be on the ptb dataset. I'm also reading the torch blog post on [NCE](torch.ch/blog/2016/07/25/nce.html)

**Comment by Stonesjtu** [comment]
The original example first concatenates all the sentences as one extremely long word sequence,  and then split them into chunks of length `bptt` **without overlay**. Thus the sentence information may flows from the previous sentence to the next one. As the sentences of PTB is extracted sequentially from the paragraphs of newspapers, the **information of previous sentence** can greatly help the LSTM to predict the next sentence.

In my experiments, the sentences are treated like independent ones. It usually gets higher PPL  ( < 10 ) than the `word_language_model` example.

NCE runs slower than pure `CrossEntropy` on PTB due to the relatively small vocabulary and high sampling overhead.

**Comment by eric-haibin-lin** [comment]
Thanks for your answers! Do you mean that if you don't treat your sentences as independent ones, but rather also "then split them into chunks of length bptt without overlay", the ppl of model trained with NCE will be better? Have you already tried that?
